ddtrace/internal/datadog/profiling/stack_v2/src/sampler.cpp

```diff
@@ -56,17 +58,63 @@ Sampler::get()
     return instance;
 }
 
+void
+_stack_v2_atfork_child()
+{
+    // The only thing we need to do at fork is to propagate the PID to echion
+    // so we don't even reveal this function to the user
+    _set_pid(getpid());
+}
+
+__attribute__((constructor)) void
+_stack_v2_init()
+{
+    _stack_v2_atfork_child();
+}
+
 void
 Sampler::one_time_setup()
 {
     _set_cpu(true);
     init_frame_cache(echion_frame_cache_size);
-    _set_pid(getpid());
+
+    // It is unlikely, but possible, that the caller has forked since application startup, but before starting echion.
+    // Run the atfork handler to ensure that we're tracking the correct process
+    _stack_v2_atfork_child();
+    pthread_atfork(nullptr, nullptr, _stack_v2_atfork_child);
 
     // Register our rendering callbacks with echion's Renderer singleton
     Renderer::get().set_renderer(renderer_ptr);
 }
 
+void
+Sampler::register_thread(uintptr_t id, uint64_t native_id, const char* name)
+{
+    const std::lock_guard<std::mutex> thread_info_guard{ thread_info_map_lock };
+
+    auto it = thread_info_map.find(id);
+    if (it == thread_info_map.end()) {
+        try {
+            thread_info_map.emplace(id, std::make_unique<ThreadInfo>(id, native_id, name));
+        } catch (const ThreadInfo::Error& e) {
+            // ignore for now
```

Comment:
Would it benefit us if we print to stderr or log?

Reply:
Honestly, I don't know.  There are a few moving parts.

1. I don't fully understand the failure surface here.  How might we fail to receive a thread ID which is also a valid pthread pointer?  I don't think this is likely--or really, possible--but if I'm wrong then these exceptions can be incredibly chatty.  The tracer goes to extreme lengths to avoid flooding stdio, and since we don't have a clear integration with the logger (so far!), we'd need to construct a new strategy for dealing with this.
2.  We can integrate with the logger more naturally downstream, but then we have a conditional in the context of a wrapper (for threading, no less), which is an uncomfortable log point for me.

Let me propose the following.
1. In the near term (this PR), we'll do what we do elsewhere in the repo, which is log to stderr once.  This will at least show that instrumentation _can_ fail and we need to be aware of it.
2. In the mid-term, I have a separate PR where I'm adding some counters to the system.  We should count the failure rate.
3. In the long-term, this is a good place to integrate with the logger probably.  I mean, it'd still be in the code path which is called by the stdlib wrapper (which I said before makes me uncomfortable), but at least it'd be in the called code instead of the top-level wrapper itself.

Reply:
I did proposal 1 in c83e1e35f

Reply:
I like your proposal for using counters.